osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

```diff
@@ -14,51 +16,97 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+                    double wideAngleBonus = calcWideAngleBonus(angle);
+                    double acuteAngleBonus = calcAcuteAngleBonus(angle);
+
+                    if (osuCurrObj.StrainTime > 100)
+                        acuteAngleBonus = 0;
+                    else
+                    {
+                        acuteAngleBonus *= Math.Min(2, Math.Pow((100 - osuCurrObj.StrainTime) / 15, 1.5));
+                        wideAngleBonus *= Math.Pow(osuCurrObj.StrainTime / 100, 6);
+                    }
+
+                    if (acuteAngleBonus > wideAngleBonus)
+                        angleBonus = Math.Min(angleBonus, 150 / osuCurrObj.StrainTime) * Math.Min(1, Math.Pow(Math.Min(osuCurrObj.JumpDistance, osuPrevObj.JumpDistance) / 150, 2));
+
+                    angleBonus *= Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier);
+
+                    // add in angle velocity.
+                    aimStrain += angleBonus;
                 }
             }
+            else // There is a rhythm change
+            {
+                // Rewarding rhythm, take the smaller velocity as base.
+                double rhythmBonus = Math.Min(currVector.Length, prevVector.Length);
 
-            double jumpDistanceExp = applyDiminishingExp(osuCurrent.JumpDistance);
-            double travelDistanceExp = applyDiminishingExp(osuCurrent.TravelDistance);
+                if (osuCurrObj.StrainTime + 10 < osuPrevObj.StrainTime && osuPrevObj.StrainTime > osuLastObj.StrainTime + 10)
+                    // Don't want to reward for a rhythm change back to back (unless its a double, which is why this only checks for fast -> slow -> fast).
+                    rhythmBonus = 0;
 
-            return Math.Max(
-                result + (jumpDistanceExp + travelDistanceExp + Math.Sqrt(travelDistanceExp * jumpDistanceExp)) / Math.Max(osuCurrent.StrainTime, timing_threshold),
-                (Math.Sqrt(travelDistanceExp * jumpDistanceExp) + jumpDistanceExp + travelDistanceExp) / osuCurrent.StrainTime
-            );
+                aimStrain += rhythmBonus * rhythm_variance_multiplier; // add in rhythm velocity.
+            }
+
+            return aimStrain;
+        }
+
+        private double calcWideAngleBonus(double angle)
+        {
+            if (angle < Math.PI / 3)
+                return 0;
+            if (angle < 2 * Math.PI / 3)
+                return Math.Pow(Math.Sin(1.5 * (angle - Math.PI / 3)), 2);
+
+            return 0.25 + 0.75 * Math.Pow(Math.Sin(1.5 * (Math.PI - angle)), 2);
         }
 
-        private double applyDiminishingExp(double val) => Math.Pow(val, 0.99);
+        private double calcAcuteAngleBonus(double angle)
+        {
+            if (angle < Math.PI / 3)
+                return 0.5 + 0.5 * Math.Pow(Math.Sin(1.5 * angle), 2);
+            if (angle < 2 * Math.PI / 3)
+                return Math.Pow(Math.Sin(1.5 * (2 * Math.PI / 3 - angle)), 2);
+
+            return 0;
+        }
```

Comment:
The wide angle bonus peaks at 120 degrees and the acute one peaks at 60, which is fine, however if they gain too much compared to the entire range of wide/acute angles it could present a problem because these particular angles buff specific patterns too much. I would try flatter curves here that also reache the peak faster, for example at 75/105 degrees.

As it currently stands the bonus difference between a 100 deg wide angle jump and a 120 deg one is about 25% which seems unrealistic to me.

Reply:
A kind user by the name of **joz** provided some [data](https://cdn.discordapp.com/attachments/886986744090734682/893153129707733082/unknown.png) on the difficulty of wide angles. This was gathered through analyzing actual replays with machine learning, and as far as I'm concerned it makes more sense than the implementation here.

The findings are that wide angles increase difficulty by up to 20%-25% regardless of bpm, and the rate of increase resembles a sigmoid function.

Reply:
If we have an implementation of this somewhere, this would likely be a great replacement to the existing wide angle calculation.